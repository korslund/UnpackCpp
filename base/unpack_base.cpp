#include "unpack_base.hpp"

#include "misc/dirwriter.hpp"
#include <mangle/stream/servers/file_stream.hpp>

using namespace Mangle::VFS;
using namespace Mangle::Stream;
using namespace UnpackCpp;

void UnpackBase::unpack(const std::string &file,
                        Mangle::VFS::StreamFactoryPtr output,
                        Progress *prog,
                        const FileList *list)
{ unpack(FileStream::Open(file),output,prog,list); }

void UnpackBase::unpack(Mangle::Stream::StreamPtr input, const std::string &dir,
                        Progress *prog,
                        const FileList *list)
{ unpack(input, StreamFactoryPtr(new UnpackCpp::DirWriter(dir)), prog, list); }

void UnpackBase::unpack(const std::string &file, const std::string &dir,
                        Progress *prog,
                        const FileList *list)
{ unpack(FileStream::Open(file),dir,prog,list); }

struct ListMaker : StreamFactory
{
  UnpackBase::FileList *list;
  StreamFactoryPtr redir;

  StreamPtr open(const std::string &name)
  {
    list->insert(name);
    if(redir)
      return redir->open(name);
    return StreamPtr();
  }
};

void UnpackBase::index(const std::string &file,
                       FileList &output,
                       Progress *prog,
                       StreamFactoryPtr write)
{ index(FileStream::Open(file),output,prog,write); }

void UnpackBase::index(StreamPtr file,
                       FileList &output,
                       Progress *prog,
                       StreamFactoryPtr write)
{
  ListMaker *m = new ListMaker;
  m->list = &output;
  m->redir = write;
  StreamFactoryPtr mp(m);

  // This basically unpacks the file normally, and pumps all the data
  // through ListMaker. The ListMaker makes note of all the filenames
  // given, and store them in 'output'.
  unpack(file, mp, prog);
}
