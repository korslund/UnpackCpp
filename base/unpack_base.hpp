#ifndef __UNPACKCPP_UNPACK_BASE_HPP
#define __UNPACKCPP_UNPACK_BASE_HPP

#include <mangle/vfs/stream_factory.hpp>
#include <set>

#include "progress.hpp"

namespace UnpackCpp
{
  /*
    Base class for all unpackers. Only the first unpack() function is
    implemented in subclasses, the others are convenient wrappers
    around it.

    The unpack() function takes an input Mangle::Stream::Stream, an
    output Mangle::VFS::StreamFactory, an optional progress updater
    functor, and an optional list of output files to produce.

    How the VFS::StreamFactory is used:

    - open("filename") is called for all files in the archive. The
      returned stream is expected to be writable, except for directory
      entries. The data for each file is written to the stream, and
      the StreamPtr is reset.

    - Returning an empty StreamPtr is allowed, and will cause the file
      to be skipped. This can be used to index the archive.

    - Filenames may include paths. The reciever is expected to create
      necessary parent directories.

    - Directories (paths ending with a slash) may be sent to
      open(). The StreamFactory is expected to create empty
      directories in those cases, and to return an empty StreamPtr.
      Both forward and backward slashes should be expected.

    - The StreamFactory itself is responsible for throwing exceptions
      on writing errors. The unpacker will only throw on unpacking
      errors.

    - Instead of using StreamFactory directly, you may use the
      directory version to write to a file system directory.

    About the FileList list:

    - If the 'list' parameter is specified, only the files listed
      there will be unpacked. Files in the list which do not match any
      in the archive, will cause an error.

      Names must match exactly with the files in the archive,
      including case and slash type. For that reason, it is recommened
      to ONLY use filenames that have previously been generated with
      the same unpacker.

    - If the list is missing or empty, all files are unpacked.

    - Implementations may choose to behave differently depending on
      whether the list is present or not. Some future EXE unpackers
      for example may enter a slow "explore mode" when a list is
      missing, but will be in a much faster "unpack mode" when given
      the right data. In those cases, the list names may contain
      additional meta-data that do not really represent actual
      filenames.
   */
  struct UnpackBase
  {
    typedef std::set<std::string> FileList;

    // Unpack the given archive file into the stream factory.
    virtual void unpack(Mangle::Stream::StreamPtr input,
                        Mangle::VFS::StreamFactoryPtr output,
                        Progress *prog = NULL,
                        const FileList *list = NULL) = 0;

    // File reader version
    void unpack(const std::string &file,
                Mangle::VFS::StreamFactoryPtr output,
                Progress *prog = NULL,
                const FileList *list = NULL);

    // Directory writer versions
    void unpack(Mangle::Stream::StreamPtr input, const std::string &dir,
                Progress *prog = NULL,
                const FileList *list = NULL);
    void unpack(const std::string &file, const std::string &dir,
                Progress *prog = NULL,
                const FileList *list = NULL);

    /* Generate an index of an archive. This is the same as running
       unpack() and gathering up the names generated through the
       StreamFactory.

       An optional StreamFactoryPtr may be provided to accept output
       data. If it is empty, no data is extracted.
    */
    void index(Mangle::Stream::StreamPtr input, FileList &output,
               Progress *prog = NULL,
               Mangle::VFS::StreamFactoryPtr write =
               Mangle::VFS::StreamFactoryPtr());
    void index(const std::string &file, FileList &output,
               Progress *prog = NULL,
               Mangle::VFS::StreamFactoryPtr write =
               Mangle::VFS::StreamFactoryPtr());
  };
}
#endif
