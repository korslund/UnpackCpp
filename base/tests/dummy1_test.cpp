#include "base/unpack_base.hpp"
#include "common.cpp"

// A dummy "unpacker" that just writes some random files
struct DummyUnpack : UnpackBase
{
  void unpack(StreamPtr input,
              StreamFactoryPtr out,
              Progress *prog = NULL,
              const FileList *list = NULL)
  {
    StreamPtr s;

    out->open("test1.dummy.empty");
    s = out->open("test2.txt");
    assert(s);
    s->write("Hello dolly\n", 12);
    s.reset();
    s = out->open("dir1/");
    //assert(!s);
    s = out->open("dir2/test3.txt");
    s->write("This is Louis, Dolly\n", 21);
  }

  using UnpackBase::unpack;
};

int main()
{
  cout << "Unpacking some fake files:\n";
  DummyUnpack dummy;

  dummy.unpack(StreamPtr(), makeFact());

  cout << "Done\n";
  return 0;
}
