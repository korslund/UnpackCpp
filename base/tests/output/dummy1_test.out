Unpacking some fake files:
FILE: test1.dummy.empty
  
FILE: test2.txt
  Hello dolly

FILE: dir1/
FILE: dir2/test3.txt
  
  This is Louis, Dolly

Done
